homeassistant/components/izone/climate.py

```diff
@@ -0,0 +1,520 @@
+        self._name = zone.name.title()
+
+        self._supported_features = (SUPPORT_ON_OFF |
+                                    SUPPORT_OPERATION_MODE)
+        if zone.type != Zone.Type.AUTO:
+            self._state_to_pizone = {
+                STATE_OFF: Zone.Mode.CLOSE,
+                STATE_FAN_ONLY: Zone.Mode.OPEN,
+            }
+        else:
+            self._state_to_pizone = {
+                STATE_OFF: Zone.Mode.CLOSE,
+                STATE_FAN_ONLY: Zone.Mode.OPEN,
+                STATE_AUTO: Zone.Mode.AUTO,
+            }
+            self._supported_features |= SUPPORT_TARGET_TEMPERATURE
+
+        self._device_info = {
+            'identifiers': {
+                (DOMAIN, controller.unique_id, zone.index)
+            },
+            'name': self.name,
+            'manufacturer': 'IZone',
+            'via_device': (DOMAIN, controller.unique_id),
+            'model': zone.type.name.title(),
+        }
+
+        device = self
+
+        # pylint: disable=protected-access
+        class ZoneListener(Listener):
+            """Listener for controller related events."""
+
+            def zone_update(
+                    self, ctrl: Controller,
+                    zone: Zone) -> None:
+                """Handle zone data updates."""
+                if zone is not device._zone:
+                    return
+                device._name = zone.name.title()
+                device.async_schedule_update_ha_state()
+
+        listener = ZoneListener()
+        controller._controller.discovery.add_listener(listener)
+        self.async_on_remove(
+            lambda: controller._controller.discovery.remove_listener(listener))
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._controller.available
+
+    @property
+    def assumed_state(self) -> bool:
+        """Return True if unable to access real state of the entity."""
+        return self._controller.assumed_state
+
+    @property
+    def device_info(self):
+        """Return the device info for the iZone system."""
+        return self._device_info
+
+    @property
+    def unique_id(self):
+        """Return the ID of the controller device."""
+        return "{}_z{}".format(
+            self._controller.unique_id, self._zone.index+1)
+
+    @property
+    def name(self) -> str:
+        """Return the name of the entity."""
+        return self._name
+
+    @property
+    def should_poll(self) -> bool:
+        """Return True if entity has to be polled for state.
+
+        False if entity pushes its state to HA.
+        """
+        return False
+
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        try:
+            if self._zone.mode == Zone.Mode.AUTO:
+                return self._supported_features
+            return self._supported_features & ~SUPPORT_TARGET_TEMPERATURE
+        except ConnectionError:
```

Comment:
Would it be better to convert the library over to pydispatcher like the zwave library does? It's my library, so there's no real constraint on how it works, but I don't want it to depend on HA directly.

Reply:
Any of the property accessors can raise a `ConnectionError` when they're not connected. They could I guess return `None` instead but I didn't like those semantics. What is the best way to approach this one?

Reply:
I would probably initialize the property to `None` and update it after the connection has been made, if that is how the flow works.

Minimally I would raise another error eg `NotConnectedError`,  better indicating what happens and what is the problem. We didn't try to connect when accessing this property, as I understand it. It's a stored exception that is raised. So this reports the state of the connection.

Reply:
You can't end up with uninitialised state, it can only go disconnected if you try to send a command and it fails. The library will keep on retrying the connection until it comes back again.

Reply:
I would still rename this error and the error message to make it clear that we're reporting the state of the connection, not that we tried to connect when accessing this property.

Reply:
Have changed it just to return `None` in this case. It's going to be flagged as unavailable anyhow.